File: leetcode/python/1_two_sum.py
"""
LeetCode Problem : Two Sum
Topic            : Array, Hash Table
Level            : Easy
URL              : https://leetcode.com/problems/two-sum
Description      : Given an array of integers nums and an integer target, return indices of the two numbers such that
    they add up to target. You may assume that each input would have exactly one solution, and you may not use the same
    element twice. You can return the answer in any order.
Examples         :
    Example 1:
    Input: nums = [2,7,11,15], target = 9
    Output: [0,1]
    Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].

    Example 2:
    Input: nums = [3,2,4], target = 6
    Output: [1,2]

    Example 3:
    Input: nums = [3,3], target = 6
    Output: [0,1]
"""

from typing import List


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        """Finds two numbers in the given list that add up to the target and returns their indices.

        Args:
            nums (List[int]): List of integers to search through.
            target (int): Target sum to find.

        Returns:
            List[int]: Indices of the two numbers that add up to target.

        Time Complexity:
            O(n): Each element is visited once in a single pass through the list.

        Space Complexity:
            O(n): In the worst case, all elements are stored in the hash table.
        """
        nums_map = {}
        for i, num in enumerate(nums):
            complement = target - num
            if complement in nums_map:
                return [nums_map[complement], i]
            nums_map[num] = i
        return []


def run_tests():
    inputs = {
        "case_1": [[2, 7, 11, 15], 9],
        "case_2": [[3, 2, 4], 6],
        "case_3": [[3, 3], 6],
    }
    outputs = {"case_1": [0, 1], "case_2": [1, 2], "case_3": [0, 1]}

    solution = Solution()

    for case, input in inputs.items():
        result = solution.twoSum(input[0], input[1])
        assert result == outputs[case], f"{case}: expected {outputs[case]}, got {result}"

    print("✅ All tests passed!")


if __name__ == "__main__":
    run_tests()

File: problems/leetcode/go/1_two_sum.go
package main

import (
	"fmt"

	"github.com/ariefsibuea/algojo/libs/go/cmp"
	"github.com/ariefsibuea/algojo/libs/go/format"
	"github.com/ariefsibuea/algojo/libs/go/runner"
)

/*
 * Problem	: Two Sum
 * Topics	: Array, Hash Table
 * Level	: Easy
 * URL		: https://leetcode.com/problems/two-sum/
 *
 * Description:
 * 		Given an array of integers nums and an integer target, return indices of the two numbers such
 * 		that they add up to target. You may assume that each input would have exactly one solution, and
 * 		you may not use the same element twice. You can return the answer in any order.
 *
 * Constraints:
 * 		- 2 <= nums.length <= 10^4
 * 		- -10^9 <= nums[i] <= 10^9
 * 		- -10^9 <= target <= 10^9
 * 		- Only one valid answer exists.
 *
 * Examples:
 * 		Example 1:
 * 		Input: nums = [2,7,11,15], target = 9
 * 		Output: [0,1]
 * 		Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].
 *
 * 		Example 2:
 * 		Input: nums = [3,2,4], target = 6
 * 		Output: [1,2]
 *
 * 		Example 3:
 * 		Input: nums = [3,3], target = 6
 * 		Output: [0,1]
 */

func twoSum(nums []int, target int) []int {
	numExists := make(map[int]int)

	for i, n := range nums {
		remaining := target - n
		if idx, ok := numExists[remaining]; ok {
			return []int{idx, i}
		}
		numExists[n] = i
	}

	return []int{}
}

func RunTestTwoSum() {
	runner.InitMetrics("TwoSum")

	testCases := map[string]struct {
		nums   []int
		target int
		expect []int
	}{
		"case-1": {
			nums:   []int{2, 7, 11, 15},
			target: 9,
			expect: []int{0, 1},
		},
		"case-2": {
			nums:   []int{3, 2, 4},
			target: 6,
			expect: []int{1, 2},
		},
		"case-3": {
			nums:   []int{3, 3},
			target: 6,
			expect: []int{0, 1},
		},
	}

	var passedCount int
	for name, tc := range testCases {
		fmt.Printf("RUN %s\n", name)
		format.PrintInput(map[string]interface{}{"nums": tc.nums, "target": tc.target})

		result := runner.ExecCountMetrics(twoSum, tc.nums, tc.target).([]int)
		if !cmp.EqualSlices(result, tc.expect) {
			format.PrintFailed("expect = %v - got = %v", tc.expect, result)
			continue
		}

		format.PrintSuccess("test case '%s' passed", name)
		passedCount++
	}

	fmt.Printf("\n📊 Test Summary: %d/%d passed\n", passedCount, len(testCases))
	runner.PrintMetrics()
}
